tests/core/policies/test_memoization.py

```diff
@@ -157,3 +157,89 @@ def test_augmented_prediction(self, max_history):
             ]
             == UTTER_BYE_ACTION
         )
+
+    @pytest.mark.parametrize("max_history", [1, 2, 3, 4, None])
+    def test_augmented_prediction_long_story(self, max_history):
```

Comment:
Thanks for these suggestions! i added additional tests and also because of them ended up finding another bug in augmemo, that is similar to but different than the one I was trying to fix:

Originally we have a test for the following situation, a training tracker that starts with an action:

```python
        training_story = TrackerWithCachedStates.from_events(
            "training story",
            [
                ActionExecuted(UTTER_GREET_ACTION),
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_3", True),
                ActionExecuted(UTTER_BYE_ACTION),
            ],
            domain=domain,
            slots=domain.slots,
        )
```

this will match the following test story because even though there are unrelated events at the start AugmentedMemoizationPolicy will keep popping initial actions off the tracker until it either exhausts the tracker or it finds a match:

```python
        test_story = TrackerWithCachedStates.from_events(
            "test story",
            [
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_1", False),
                ActionExecuted(UTTER_GREET_ACTION),
                ActionExecuted(UTTER_GREET_ACTION),
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_2", True),
                ActionExecuted(UTTER_GREET_ACTION),
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_3", True),
                # ActionExecuted(UTTER_BYE_ACTION),
            ],
            domain=domain,
            slots=domain.slots,
        )
```        
        
however if the training story starts with a UserUttered event it will not find a match, e.g.:

```python
        training_story = TrackerWithCachedStates.from_events(
            "training story",
            [
                UserUttered(intent={"name": GREET_INTENT_NAME}),
                ActionExecuted(UTTER_GREET_ACTION),
                SlotSet("slot_3", True),
                ActionExecuted(UTTER_BYE_ACTION),
            ],
            domain=domain,
            slots=domain.slots,
        )
```
because the test story will get truncated to after the UserUttered event, which change the underlying states. 

Not that this is not about tracker truncation which happens [here](https://github.com/RasaHQ/rasa/blob/a9d885f20add8f2bdcc5b65b6863555dd68c9fc5/rasa/core/policies/memoization.py#L401) and that I had fixed. This new issue happens [here](https://github.com/RasaHQ/rasa/blob/a9d885f20add8f2bdcc5b65b6863555dd68c9fc5/rasa/core/policies/memoization.py#L290) in the `_back_to_the_future` method. This method originally tried to truncate the tracker at the next action, effectively removing user utterance information from the state.  I was able to fix it by making the `_back_to_the_future` method truncate to the next action **or user uttered event.** This will add extra checks but since its only user uttereances and only one of those can happen per arbitrarily many bot actions I think it will be negligible.

Reply:
@kedz I admit that I haven't looked at the code (following your links) but I've verified that, even if a training story starts with an intent, it gets translated into a tracker (and subsequently into states) where there's `action_listen` before the intent. So, I think the scenario you described can't happen today. In any case, AugMemo has lots of "magical" code that is loaded with assumptions and we could at least make sure this one (stories starting with actions) is made clear in there (or is even checked for?). Wdyt?